### PR TITLE
Add await interval

### PR DESCRIPTION
**Pull request by bernardd**
I'm using Chaperon to do some load testing on our Absinthe (https://github.com/absinthe-graphql/absinthe) GraphQL server. The issue I've run into is that, when running with a phoenix websocket, I need to send a heartbeat every 30 seconds or so while idling or the server drops the connection. Unfortunately that doesn't fit well with the model used by chaperon where stuff happens basically in a strict order, and so calling the blocking `await` or `await_signal` functions doesn't allow for a purely time-based number of heatbeat send/receives during the waiting.

This change attempts to provide a generic method by which this can be done, adding an `interval` config option of the form `{period, function}`. The effect is that during any `await` operation, `function` is called once per `period` milliseconds. In this way we can inject a `ws_send`/`ws_recv` pair to handle the heartbeat and response while still waiting.

I've implemented it for both `await_signal` and `await` (task) however there's currently a bit of a difference: The first variant uses the main worker process and a single session, while the latter spawns off a separate worker process and merges the session back at the end of the wait period. I'm not especially happy with that discrepancy, so I'm more than happy to take feedback on that (or anything else) to make it better.

As part of this I also wrote a bunch of tests for the existing `await` functionality to make sure I didn't break it too badly.

### CODE REVIEW

**Comment by bakkdoor** [comment]
Thanks for this contribution! It looks pretty good to me from glancing over it but I will do some additional testing and let you know if I find anything that might need some changes before I can merge it. Will get back to you soon. 👍

**Comment by bernardd** [comment]
No worries - it's definitely a non-trivial change (hence all the tests I added) so I won't be offended in the slightest if you feel like there's changes needed :)

**Review by bakkdoor — APPROVED** [review]
OK, going to merge this. I have a couple of auto-format / code style fixes in a commit I will merge after this. 👍

**Pull request merged**
